src/main/java/org/opensearch/security/user/User.java

```diff
@@ -272,15 +274,16 @@ public synchronized final Map<String, String> getCustomAttributesMap() {
         return attributes;
     }
 
-    public final void addSecurityRoles(final Collection<String> securityRoles) {
-        if (securityRoles != null && this.securityRoles != null) {
-            this.securityRoles.addAll(securityRoles);
+    public final void setSecurityRoles(final Collection<String> securityRoles) {
```

Comment:
This perpetuates a confusing design, its still a side-effect.  Remove this field entirely from User, and create a subclass UserWithRoles that extends User.  UserWithRoles could require the mapped roles as part of its construction ensuring they aren't modified after the user has been created.

Then in places where `user.getSecurityRoles()` is called we can work backwards on what needs to be fixed so the UserWithRoles object is chained correctly ensuring a single assignment and no possiblity of dynamic update.

IF we see that dynamic updates are taking place we can address them individually